### ISSUE TITLE
`readability-const-return-type` warning lacks context

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#73270]
```cpp
#include <algorithm>
#include <map>
#include <set>

struct SelectMapKeys
{
    template<class Pair>
    typename Pair::first_type operator()(const Pair& p) const
    {
        return p.first;
    }
};

void f(std::map<int, int> s)
{
    std::set<int> varids;
    std::transform(s.begin(), s.end(), std::inserter(varids, varids.begin()), SelectMapKeys{});
}
```

```
<source>:8:5: warning: return type 'typename pair<const int, int>::first_type' (aka 'const int') is 'const'-qualified at the top level, which may reduce code readability without improving const correctness [readability-const-return-type]
    8 |     typename Pair::first_type operator()(const Pair& p) const
      |     ^
```

https://godbolt.org/z/aMMWfGvTK

In the reduced code it is obvious since there is only one usage. In the original code the struct is in a header with no indication of the actual usage which is causing this.

I also think this might be a false positive - but that needs to be handled in a different ticket.

**Comment by PiotrZSL** [comment]
Yes thats false-positive.

**Comment by firewave** [comment]
This does fix the false positive but not the case of having the missing context information. I will try to provide another example with a true positive.

**Comment by SwuduSusuwu** [comment]
With const return, compiler catches this:
```
const int f();
int main() {
assert(f() = 0) ;
}
```

If mutable, this compiles:
```
int f();
int main() {
assert(f() = 0) ;
}
```
Suggest to remove "readability-const-return-type", as const correctness secures us.
